Objects/unicodeobject.c

```diff
@@ -291,12 +291,18 @@ hashtable_unicode_compare(const void *key1, const void *key2)
    It's not safe to deallocate those strings until all interpreters that
    potentially use them are freed.  By storing them in the main interpreter, we
    ensure they get freed after all other interpreters are freed.
+
+   Subtle detail: it's only required to share the interned string dict in the
+   case that those kinds of legacy modules are actually imported.  However, we
+   can't wait until the import happens so we share if those kind of modules are
+   allowed (the Py_RTFLAGS_MULTI_INTERP_EXTENSIONS flag is set).
 */
 static bool
 has_shared_intern_dict(PyInterpreterState *interp)
 {
     PyInterpreterState *main_interp = _PyInterpreterState_Main();
-    return interp != main_interp  && interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC;
+    return (interp != main_interp  &&
+            !(interp->feature_flags & Py_RTFLAGS_MULTI_INTERP_EXTENSIONS));
```

Comment:
FYI, (not `Py_RTFLAGS_MULTI_INTERP_EXTENSIONS`) does not imply `Py_RTFLAGS_USE_MAIN_OBMALLOC`.  Why not stick with `Py_RTFLAGS_USE_MAIN_OBMALLOC`?

Reply:
I'm looking at `reload_singlephase_extension` and when we can end up in the `PyDict_Update` case.  I think it can only happen if `Py_RTFLAGS_MULTI_INTERP_EXTENSIONS` is set.  If we do that dict update then we are sharing objects between interpreters.